component/src/shielded_pool/component.rs

```diff
@@ -541,33 +764,178 @@ pub trait View: StateExt {
+            // been rolled back and we don't have to care about it here:
+            if !source.is_empty() {
+                // Non-empty source means we should be able to decode it
+                let source = NoteSource::decode(&*source)?;
+
+                return Err(anyhow!(
+                    "nullifier {} was already spent in {:?} (currently quarantined)",
+                    nullifier,
+                    source,
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn quarantined_to_apply(&self, epoch: Epoch) -> Result<Quarantined> {
+        Ok(self
+            .get_domain(state_key::quarantined_to_apply(epoch.index))
+            .await?
+            .unwrap_or_default())
+    }
+
+    async fn schedule_unquarantine(&self, epoch: Epoch, quarantined: Quarantined) -> Result<()> {
+        let mut updated_quarantined = self.quarantined_to_apply(epoch).await?;
+        updated_quarantined.extend(quarantined);
+        self.put_domain(
+            state_key::quarantined_to_apply(epoch.index),
+            updated_quarantined,
+        )
+        .await;
+        Ok(())
+    }
+
+    async fn unschedule_unquarantine(
+        &self,
+        epoch: Epoch,
+        identity_key: IdentityKey,
+    ) -> Result<super::quarantined::PerValidator> {
+        let mut updated_quarantined = self.quarantined_to_apply(epoch).await?;
+        let per_validator = updated_quarantined
+            .quarantined
+            .remove(&identity_key)
+            .unwrap_or_default();
+        // We're removed all the scheduled notes and nullifiers for this epoch and identity key:
+        self.put_domain(
+            state_key::quarantined_to_apply(epoch.index),
+            updated_quarantined,
+        )
+        .await;
+        // Now we also ought to remove these nullifiers and notes from quarantine, but *not* apply them:
+        for &nullifier in per_validator.nullifiers.iter() {
+            self.unquarantine_nullifier(nullifier).await?;
+        }
+        for note_payload in per_validator.notes.iter() {
+            self.unquarantine_note(&note_payload.note_commitment)
+                .await?;
+        }
+
+        // Return the per_validator structure so we can use it later
+        Ok(per_validator)
+    }
+
+    // Unschedule the unquarantining of all notes and nullifiers for the given validator, in any
+    // epoch which could possibly still be unbonding
+    async fn unschedule_unquarantine_all(&self, identity_key: IdentityKey) -> Result<()> {
+        let height = self.get_block_height().await?;
+        let epoch_duration = self.get_epoch_duration().await?;
+        let this_epoch = Epoch::from_height(height, epoch_duration);
+        let unbonding_epochs = self.get_chain_params().await?.unbonding_epochs;
+
+        // We will edit the compact block to note the rolled back nullifiers and commitments
+        let mut compact_block = self.compact_block(height).await?.unwrap_or_default();
```

Comment:
We cannot edit the compact block like this, because it's stored in memory in the shielded pool component, not in the state.  This method needs to be a method on the shielded pool component, because that's the only thing that has access to the compact block.

Reply:
If we want the staking component to trigger changes to the compact block, we need to use the state to do message-passing to the shielded component, which owns it.